force-app/main/default/classes/DMLOperations.cls

```diff
@@ -145,11 +173,63 @@ public with sharing class DMLOperations {
      * @param oppNames The list of Opportunity names.
      */
     public static void upsertOpportunities(String accountName, List<String> oppNames) {
+        // Create an empty list to hold Opportunity records to be added
+        List<Opportunity> oppsToAdd = new List<Opportunity>();
+    
+        // Query to find the Account with the specified name
+        Account myAcc;
+        List<Account> accountList = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
+    
+        // Check if the account was found
+        if (!accountList.isEmpty()) {
+            myAcc = accountList[0];
+        } else {
+            // If no account is found, create a new Account with the given name
+            myAcc = new Account(Name = accountName);
+            // Upsert the new Account (insert if new, update if existing)
+            upsert myAcc;
+        }
+    
+        // Query for existing Opportunities with the same name under the current account
+        Set<String> existingOppNames = new Set<String>();
+        List<Opportunity> existingOpps = [
+            SELECT Name 
+            FROM Opportunity 
+            WHERE AccountId = :myAcc.Id 
+            AND Name IN :oppNames
+        ];
+    
+        // Add existing opportunity names to the set
+        for (Opportunity opp : existingOpps) {
+            existingOppNames.add(opp.Name);
+        }
+    
         // Loop through each Opportunity name
-            // Create an Opportunity with the given name
-            // Add the Opportunity to the list
+        for (String oppName : oppNames) {
+            // Check if the Opportunity already exists
+            if (!existingOppNames.contains(oppName)) {
+                // Create a new Opportunity with the given name if it doesn't already exist
+                Opportunity newOpp = new Opportunity();
+                newOpp.StageName = 'Qualification'; // Set a default Stage Name
+                newOpp.CloseDate = Date.today().addDays(5); // Set a default Close Date
+                newOpp.Name = oppName; 
+                newOpp.AccountId = myAcc.Id; 
+    
+                // Add the Opportunity to the list for upserting
+                oppsToAdd.add(newOpp);
+            }
```

Comment:
What happens if the existing opp is found, should the values still be applied for upsert? Otherwise good job avoiding DML inside a loop.